### PR TITLE
annotated

### PR DESCRIPTION
**Pull request by fametrano**
As exercise to explore, understand, and possibly extend hwi, I have annotated the HardwareWalletClient and moved code comments as proper doc-strings. Of course I might have been missing something, so suggestions/corrections are welcome and I will update the annotations accordingly.

Being the main class, HardwareWalletClient is the first file a developer would look at and IMO deserves to be annotated to easily explain the class intended usage.

If this contribution is appreciated I am willing to annotate more (eventually all) code, on this PR branch or using a separate follow up PR.

### CODE REVIEW

File: hwilib/hwwclient.py [hunk 1]
```diff
@@ -1,82 +1,167 @@
...
+
+        Return {"xpub": xpub}
+        """
+        # FIXME is "m/44'/0'/0'" just an example?
+        # how is testnet handled?
+        # shouldn't this be
+        # return self.get_pubkey_at_path(self.path)
+        # or according to get_master_fingerprint_hex
+        # return self.get_pubkey_at_path("m/0h")
+        return self.get_pubkey_at_path("m/44'/0'/0'")
```
Comment (achow101) [inline hwilib/hwwclient.py:34]: Re your questions: This follows the BIP 44 default, although it indeed should be different for testnet.

File: hwilib/hwwclient.py [hunk 2]
```diff
@@ -1,82 +1,167 @@
...
+    def get_pubkey_at_path(self, bip32_path: str) -> Dict[str, str]:
+        """Return a dictionary with the xpub.
+
+        Retrieve the public key at the specified BIP32 derivation path.
+
+        Return {"xpub": xpub}
+        """
+        raise NotImplementedError("not implemented by HardwareWalletClient base class")
+
+    def sign_tx(self, tx) -> str:
```
Comment (achow101) [inline hwilib/hwwclient.py:50]: This actually returns a `Dict[str, str]`. All of these functions should return a dict.

File: hwilib/hwwclient.py [hunk 3]
```diff
@@ -1,82 +1,167 @@
...
+        """Return a hex string with the signed transaction.
```
Comment (achow101) [inline hwilib/hwwclient.py:51]: The return value is a base64 encoded PSBT, not hex. The original comment is out of date.

it returns a dict of the form

    {'psbt': <base64 psbt string>}

File: hwilib/hwwclient.py [hunk 4]
```diff
@@ -1,82 +1,167 @@
...
+
+        The tx must be in the combined unsigned transaction format.
```
Comment (achow101) [inline hwilib/hwwclient.py:53]: The tx is a `PSBT` object (see `serializations.py`)

File: hwilib/hwwclient.py [hunk 5]
```diff
@@ -1,82 +1,167 @@
...
+        """
+        raise NotImplementedError("not implemented by HardwareWalletClient base class")
+
+    def sign_message(self, message: str, bip32_path: str) -> str:
+        """Return a base64 encoded string with the signed message.
```
Comment (achow101) [inline hwilib/hwwclient.py:58]: This returns a dict of the form

    {'signature': <base64 signature string>}

File: hwilib/hwwclient.py [hunk 6]
```diff
@@ -1,82 +1,167 @@
...
-    def toggle_passphrase(self):
-        raise NotImplementedError('The HardwareWalletClient base class does not implement this method')
+    def get_pubkey_at_path(self, bip32_path: str) -> Dict[str, str]:
+        """Return a dictionary with the xpub.
+
+        Retrieve the public key at the specified BIP32 derivation path.
+
+        Return {"xpub": xpub}
+        """
+        raise NotImplementedError("not implemented by HardwareWalletClient base class")
```
Comment (achow101) [inline hwilib/hwwclient.py:48]: We should avoid changing these `raise` lines to preserve the git blame. The error message should also remain as it was.

**Comment by fametrano** [comment]
Hopefully I've addressed all points and I look forward to other comments/suggestions

**Comment by achow101** [comment]
ACK, please squash.

**Comment by fametrano** [comment]
done

**Comment by achow101** [comment]
ACK 465f4fd8ed1800cb42197cdff1058af00511ad83

**Pull request merged**